src/server/render.js

```diff
@@ -2,80 +2,47 @@
 // to the client
 
 import React from "react";
+import Router, {DefaultRoute, Link, Route, RouteHandler} from 'react-router';
+import routes from "../routes";
 import serialize from "serialize-javascript";
 
-import app from "../app";
 import HtmlDocument from "./HtmlDocument";
 
-import { navigateAction } from "fluxible-router";
-
 let webpackStats;
 
 if (process.env.NODE_ENV === "production") {
     webpackStats = require("./webpack-stats.json");
 }
 
-function renderApp(req, res, context, next) {
+function render(req, res, next) {
     try {
 
+        // In development, reload webpack stats file on every request
         if (process.env.NODE_ENV === "development") {
             webpackStats = require("./webpack-stats.json");
-
-            // Do not cache webpack stats: the script file would change since
-            // hot module replacement is enabled in the development env
             delete require.cache[require.resolve("./webpack-stats.json")];
         }
 
-        // dehydrate the app and expose its state
-        const state = "window.App=" + serialize(app.dehydrate(context)) + ";";
-
-        const Application = app.getComponent();
-
-        // Render the Application to string
-        const markup = React.renderToString(
-          <Application context={ context.getComponentContext() } />
-        );
+        Router.run(routes, req.path, function routeMatched(Root, state) {
+            const markup = React.renderToString(<Root/>);
+
+            // The application component is rendered to static markup
+            // and sent as response.
+            const html = React.renderToStaticMarkup(
+              <HtmlDocument
+                markup={markup}
+                script={webpackStats.script}
+                css={webpackStats.css}
+              />
+            );
+            const doctype = "<!DOCTYPE html>";
+            res.send(doctype + html);
+        });
```

Comment:
Once on the server (when a request comes in) and once on the client (on `window.history.push`)